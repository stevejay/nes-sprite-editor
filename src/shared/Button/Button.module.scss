@import "styles/_base.scss";

.button {
  border: 0px solid;
  padding: rem(12) rem(24);
  margin: 0;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  -webkit-appearance: none;
  background-clip: padding-box;
  cursor: pointer;
  display: inline-block;
  border: none;
  border-radius: rem(3);
  font-style: normal;
  text-align: center;
  text-decoration: none;
  user-select: none;
  min-height: rem(39);
  font-size: rem(16);
  color: $md-white;
  background-color: $md-light-green-800;
  will-change: background-color;
  transition: background-color $transition-fast;

  &::-moz-focus-inner {
    border: 0; /* Removes dotted line in Firefox */
  }

  &:hover {
    background-color: darken($md-light-green-800, 5);
  }
}

.button:focus {
  // outline: $focus-outline;
  @include focus-ring;
}

// // http://browserhacks.com/#hack-f4ade0540d8e891e8190065f75acb186 for webkit browsers
// .button:focus:not(*:root) {
//   outline: $webkit-focus-outline;
// }
