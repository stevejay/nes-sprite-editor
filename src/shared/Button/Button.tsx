import React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";
import { isNil } from "lodash";
import { IconType } from "react-icons/lib/cjs";

export type Props = {
  type?: "button" | "submit";
  role?: string;
  icon?: IconType;
  children?: React.ReactNode;
  tabIndex?: number;
  disabled?: boolean;
  ["aria-label"]?: string;
  size?: "small" | "medium";
  appearance?: "default" | "dark" | "primary" | "input" | "transparent";
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLButtonElement>) => void;
};

const Button = React.forwardRef<HTMLButtonElement, Props>(
  (
    {
      type = "button",
      role,
      icon,
      children,
      tabIndex = 0,
      disabled = false,
      size = "medium",
      appearance = "default",
      className,
      onClick,
      onKeyDown,
      ...rest
    }: Props,
    ref
  ) => {
    const buttonClassNames = classNames(
      styles.button,
      styles[size],
      styles[appearance],
      className
    );

    return (
      <button
        {...rest}
        ref={ref}
        type={type}
        role={role}
        disabled={disabled}
        className={buttonClassNames}
        tabIndex={tabIndex}
        onClick={onClick}
        onKeyDown={onKeyDown}
      >
        {icon && React.createElement(icon)}
        {!isNil(children) && <span>{children}</span>}
      </button>
    );
  }
);

export default Button;
