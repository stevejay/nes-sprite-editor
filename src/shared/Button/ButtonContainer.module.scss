@import "styles/_base.scss";

.container {
  margin-bottom: 1rem;
}

.container button {
  margin-right: 1rem;

  &:last-of-type {
    margin-right: 0;
  }
}
