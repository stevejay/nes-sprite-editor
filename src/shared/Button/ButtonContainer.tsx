import React from "react";
import styles from "./ButtonContainer.module.scss";

type Props = {
  children: React.ReactNode;
};

const ButtonContainer = ({ children }: Props) => (
  <div className={styles.container}>{children}</div>
);

export default ButtonContainer;
