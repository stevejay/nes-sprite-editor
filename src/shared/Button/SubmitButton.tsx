import React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";

type Props = {
  label: string;
  tabIndex?: number;
  disabled?: boolean;
  ariaLabel?: string;
  size?: "small" | "medium";
  color?: "default" | "primary" | "transparent";
  className?: string;
};

const SubmitButton = ({
  label,
  tabIndex = 0,
  disabled = false,
  ariaLabel,
  size = "medium",
  color = "default",
  className
}: Props) => {
  const buttonClassNames = classNames(
    styles.button,
    styles[size],
    styles[color],
    className
  );

  return (
    <input
      type="submit"
      disabled={disabled}
      className={buttonClassNames}
      tabIndex={tabIndex}
      aria-label={ariaLabel}
      value={label}
    />
  );
};

export default SubmitButton;
