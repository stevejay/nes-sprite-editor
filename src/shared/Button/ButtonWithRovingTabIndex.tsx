import React from "react";
import Button, { Props as ButtonProps } from "./Button";
import { useRovingTabIndex } from "../RovingTabIndex";
import useFocusEffect from "../utils/use-focus-effect";

type Props = {
  type?: ButtonProps["type"];
  icon?: ButtonProps["icon"];
  children?: ButtonProps["children"];
  disabled?: ButtonProps["disabled"];
  ariaLabel?: ButtonProps["ariaLabel"];
  size?: ButtonProps["size"];
  color?: ButtonProps["color"];
  className?: ButtonProps["className"];
  onClick?: ButtonProps["onClick"];
  onKeyDown?: ButtonProps["onKeyDown"];
};

const ButtonWithRovingTabIndex = ({
  disabled = false,
  onClick: parentOnClick,
  onKeyDown: parentOnKeyDown,
  ...rest
}: Props) => {
  const buttonRef = React.useRef<HTMLButtonElement>(null);

  const [tabIndex, focused, onKeyDown, onClick] = useRovingTabIndex(
    buttonRef,
    disabled
  );

  useFocusEffect(focused, buttonRef);

  return (
    <Button
      {...rest}
      ref={buttonRef}
      disabled={disabled}
      tabIndex={tabIndex}
      onClick={(event: React.MouseEvent<any>) => {
        parentOnClick && parentOnClick(event);
        onClick();
      }}
      onKeyDown={(event: React.KeyboardEvent<any>) => {
        parentOnKeyDown && parentOnKeyDown(event);
        onKeyDown(event);
      }}
    />
  );
};

export default ButtonWithRovingTabIndex;
