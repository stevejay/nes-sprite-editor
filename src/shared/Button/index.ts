import Button from "./Button";
import SubmitButton from "./SubmitButton";
import ButtonContainer from "./ButtonContainer";

type ButtonFn = typeof Button & {
  Container: typeof ButtonContainer;
  Submit: typeof SubmitButton;
};

const ButtonExport = Button as ButtonFn;
ButtonExport.Container = ButtonContainer;
ButtonExport.Submit = SubmitButton;

export default ButtonExport;
