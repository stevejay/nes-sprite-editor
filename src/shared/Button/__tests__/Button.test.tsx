import React from "react";
import { noop } from "lodash";
import { render } from "@testing-library/react";
import Button from "../Button";
import { FiX } from "react-icons/fi";

test("displays correctly with icon", async () => {
  const { container } = render(<Button icon={FiX} onClick={noop} />);
  expect(container).toMatchSnapshot();
});
