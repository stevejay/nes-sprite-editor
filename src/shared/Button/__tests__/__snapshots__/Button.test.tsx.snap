// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correctly with icon 1`] = `
<div>
  <button
    class="button medium default"
    tabindex="0"
    type="button"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="18"
        x2="6"
        y1="6"
        y2="18"
      />
      <line
        x1="6"
        x2="18"
        y1="6"
        y2="18"
      />
    </svg>
  </button>
</div>
`;
